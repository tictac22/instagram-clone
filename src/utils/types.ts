export interface EventWithValue extends Event {
	value: string
}

export type Author = {
	uid: string
	fullName: string
	photoUrl: string
	userName: string
}
export type Post = {
	text: string
	id: string
	uid: string
	images: string[]
	author: Author
	createdAt: {
		seconds: number
	}
}
