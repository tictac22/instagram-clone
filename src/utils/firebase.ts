// Import the functions you need from the SDKs you need
import { initializeApp } from "firebase/app"

import {
	createUserWithEmailAndPassword,
	deleteUser,
	FacebookAuthProvider,
	getAuth,
	signInWithPopup
} from "firebase/auth"
import {
	collection,
	doc,
	getDocs,
	getFirestore,
	query,
	setDoc,
	where
} from "firebase/firestore"
const firebaseConfig = {
	apiKey: import.meta.env.VITE_apiKey,
	authDomain: import.meta.env.VITE_authDomain,
	projectId: import.meta.env.VITE_projectId,
	storageBucket: import.meta.env.VITE_storageBucket,
	messagingSenderId: import.meta.env.VITE_messagingSenderId,
	appId: import.meta.env.VITE_appId,
	measurementId: import.meta.env.VITE_measurementId
}

// Initialize Firebase
const app = initializeApp(firebaseConfig)
export const db = getFirestore(app)
export const auth = getAuth()

type SignUp = {
	email: string
	password: string
	username: string
	fullName: string
}

export const signInWithFacebook = async () => {
	try {
		const provider = new FacebookAuthProvider()
		const result = await signInWithPopup(auth, provider)

		const usersCollectionRef = doc(db, "users", result.user.uid)
		await setDoc(usersCollectionRef, {
			userId: result.user.uid
		})
	} catch (error) {
		console.log(error)
		throw new CustomError(ErrorMessages.email, "email")
	}
}
export const signUp = async ({
	email,
	password,
	username,
	fullName
}: SignUp) => {
	try {
		const registration = await createUserWithEmailAndPassword(
			auth,
			email,
			password
		)
		const isTaken = await isNameTaken(username)
		if (isTaken) {
			await deleteUser(registration.user)
			throw new CustomError(ErrorMessages.username, "username")
		}
		const usersCollectionRef = doc(db, "users", registration.user.uid)
		await setDoc(usersCollectionRef, {
			userId: registration.user.uid,
			username,
			fullName
		})
	} catch (error) {
		if (error instanceof CustomError) {
			throw error
		}
		throw new CustomError(ErrorMessages.email, "email")
	}
}
const isNameTaken = async (username: string) => {
	try {
		const usersCollectionRef = collection(db, "users")
		const q = query(usersCollectionRef, where("username", "==", username))

		const querySnapshot = await getDocs(q)

		const users = []
		querySnapshot.forEach((item) => {
			users.push({ ...item.data(), id: item.id })
		})
		if (users.length >= 1) return true
		return false
	} catch (error) {
		console.error(error)
	}
}

class CustomError extends Error {
	status: number
	data: object
	constructor(message: string, field: string) {
		super(message)
		this.status = 404
		this.data = {
			[field]: [message]
		}
	}
}

const enum ErrorMessages {
	email = "Email is already in use",
	username = "Username is already in use"
}
