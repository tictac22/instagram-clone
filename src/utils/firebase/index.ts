export * from "./auth"
export * from "./config"
export * from "./post"
export * from "./storage"
