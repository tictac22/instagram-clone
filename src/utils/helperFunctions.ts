// @ts-nocheck
export function debounce(fn: () => void, delay: number) {
	let timeoutID = null
	return function () {
		clearTimeout(timeoutID)
		const args = arguments
		const that = this
		timeoutID = setTimeout(function () {
			fn.apply(that, args)
		}, delay)
	}
}

export const timeAgo = (input: number) => {
	const date = new Date(input * 1000)
	const formatter = new Intl.RelativeTimeFormat("en")
	const ranges = {
		years: 3600 * 24 * 365,
		months: 3600 * 24 * 30,
		weeks: 3600 * 24 * 7,
		days: 3600 * 24,
		hours: 3600,
		minutes: 60,
		seconds: 1
	}
	const secondsElapsed = (date.getTime() - Date.now()) / 1000
	for (const key in ranges) {
		if (ranges[key] < Math.abs(secondsElapsed)) {
			const delta = secondsElapsed / ranges[key]
			return formatter.format(Math.round(delta), key)
		}
	}
}
