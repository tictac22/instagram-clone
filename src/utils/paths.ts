export const enum RoutePaths {
	HOME = "/",
	LOGIN = "/register/login",
	SIGNUP = "/register/signup",
	POST = "/p/:id",
	USER = "/:id"
}
