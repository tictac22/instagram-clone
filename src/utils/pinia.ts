import { defineStore } from "pinia"
import { reactive } from "vue"

type User = {
	userName: string
	fullName: string
	uid: string
	photoUrl: string
	subscribed: string[]
	likes: string[]
}
export const useUserStore = defineStore("counter", () => {
	const user = reactive<{ isAuthenticated: null | boolean; data: User }>({
		isAuthenticated: null,
		data: {
			userName: "",
			fullName: "",
			uid: "",
			photoUrl: "",
			subscribed: [],
			likes: []
		}
	})

	const handleLike = async (id: string) => {
		const isLiked = user.data.likes.find(item => item === id)
		if (isLiked) {
			const index = user.data.likes.indexOf(isLiked)
			user.data.likes.splice(index, 1)
		} else {
			user.data.likes.push(id)
		}
		const { likePost } = await import("@/utils/firebase")
		likePost(user.data.uid, user.data.likes)
	}

	const setUser = (isAuthenticated: boolean, userData: User) => {
		user.isAuthenticated = isAuthenticated
		user.data = userData
	}
	const authenficate = async () => {
		const { auth, getUser } = await import("./firebase")
		const { onAuthStateChanged } = await import("firebase/auth")

		onAuthStateChanged(auth, async firebaseUser => {
			await delay(200)
			if (!firebaseUser) return (user.isAuthenticated = false)
			if (firebaseUser?.providerData[0].providerId === "facebook.com") {
				const userData = await getUser(firebaseUser!.uid)!
				if (!userData) {
					user.data = {
						uid: firebaseUser!.uid,
						...firebaseUser!
					}
					return (user.isAuthenticated = false)
				}
				user.data = {
					uid: firebaseUser!.uid,
					...userData
				}
				return (user.isAuthenticated = true)
			}
			const userData = await getUser(firebaseUser!.uid)!
			user.data = {
				uid: firebaseUser!.uid,
				...userData
			}
			user.isAuthenticated = true
		})
	}
	return { user, authenficate, setUser, handleLike }
})

const delay = (time: number) =>
	new Promise(resolve => {
		setTimeout(resolve, time)
	})
