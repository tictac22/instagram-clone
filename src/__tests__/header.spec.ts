import { beforeEach, describe, expect, it, vi } from "vitest"

import Header from "@/components/header/header.vue"
import { useUserStore } from "@/utils/pinia"
import { mount, VueWrapper } from "@vue/test-utils"
import { createPinia, setActivePinia } from "pinia"

type store = ReturnType<typeof useUserStore>
const mockRoute = {
	path: ""
}
vi.mock("@/utils/pinia", () => ({
	useUserStore: vi.fn(() => ({
		user: {
			isAuthenticated: false
		}
	}))
}))
describe("Header component", () => {
	let wrapper: VueWrapper
	let store: store
	beforeEach(() => {
		wrapper = mount(Header, {
			shallow: true,
			global: {
				stubs: ["FontAwesomeIcon"],
				mocks: {
					route: mockRoute
				}
			}
		})
		setActivePinia(createPinia())
		store = useUserStore()
	})
	it("mounts ssuccesffulyy with correct pinia data", () => {
		expect(wrapper.text()).toContain("Sign up")
		expect(store.user.isAuthenticated).toEqual(false)
		expect(wrapper.element).toMatchSnapshot()
	})
})
