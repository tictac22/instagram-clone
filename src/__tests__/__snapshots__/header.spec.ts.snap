// Vitest Snapshot v1

exports[`Header component > mounts ssuccesffulyy with correct pinia data 1`] = `
<div
  data-v-app=""
>
  
  <header
    class="sticky top-0 z-20 border-b border-solid border-[#DBDBDB] bg-white"
  >
    <div
      class="m-auto flex h-[60px] max-w-5xl items-center justify-between px-5"
    >
      <div>
        <routerlink
          to="/"
        >
          <instagram-logo-stub />
        </routerlink>
      </div>
      <search-stub />
      <div
        class="flex items-center"
      >
        <div>
          <routerlink
            class="inline-block rounded bg-blue-500 px-3 py-1 uppercase text-white"
            to="/register/login"
          >
             Log in 
          </routerlink>
          <routerlink
            class="inline-block rounded px-3 py-1 uppercase text-blue-500"
            to="/register/signup"
          >
             Sign up 
          </routerlink>
        </div>
      </div>
    </div>
  </header>
  <loading-bar-stub />
  
</div>
`;
